src/gridcoin/voting/poll.cpp

```diff
@@ -224,16 +300,133 @@ std::string Poll::WeightTypeToString() const
+    case PollResponseType::YES_NO_ABSTAIN:  return _("Yes/No/Abstain");
+    case PollResponseType::SINGLE_CHOICE:   return _("Single Choice");
+    case PollResponseType::MULTIPLE_CHOICE: return _("Multiple Choice");
+    case PollResponseType::OUT_OF_BOUND:    break;
     }
 
     assert(false); // Suppress warning
 }
 
+const std::vector<Poll::PollTypeRules> Poll::POLL_TYPE_RULES = {
+    // These must be kept in the order that corresponds to the PollType enum.
+    // { min duration, min vote percent AVW, { vector of required additional fieldnames } }
+    {  0,  0, {} },                                // PollType::UNKNOWN
+    // Note there is NO payload version protection on the vector of required additional fieldnames
+    // and all payloads less than v3 only allowed PollType::SURVEY. Furthermore, the serialization
+    // of the poll class additional fields depends only on whether the poll type is SURVEY. The net
+    // of this is that the required additional fieldnames need to remain an empty vector for SURVEY.
+    //
+    // If a new SURVEY type is needed in the future with additional fields, a new enum entry should
+    // be created for it.
+    //
+    // In addition note that any poll type that has a min vote percent AVW requirement must
+    // also require the weight type of BALANCE_AND_MAGNITUDE, so therefore the
+    // only poll type that can actually use BALANCE is SURVEY. All other WeightTypes are deprecated.
+    {  7,  0, {} },                                // PollType::SURVEY
+    { 21, 40, { "project_name", "project_url" } }, // PollType::PROJECT
+    { 42, 50, {} },                                // PollType::DEVELOPMENT
+    { 21, 20, {} },                                // PollType::GOVERNANCE
+    { 21, 40, {} },                                // PollType::MARKETING
+    { 21, 40, {} },                                // PollType::OUTREACH
+    { 21, 10, {} }                                 // PollType::COMMUNITY
+};
+
+// -----------------------------------------------------------------------------
+// Class: Poll::AdditionalFieldList
+// -----------------------------------------------------------------------------
+using AdditionalFieldList = Poll::AdditionalFieldList;
+using AdditionalField = Poll::AdditionalField;
+
+AdditionalFieldList::AdditionalFieldList(std::vector<AdditionalField> additional_fields)
+    : m_additional_fields(std::move(additional_fields))
+{
+}
+
+AdditionalFieldList::const_iterator AdditionalFieldList::begin() const
+{
+    return m_additional_fields.begin();
+}
+
+AdditionalFieldList::const_iterator AdditionalFieldList::end() const
+{
+    return m_additional_fields.end();
+}
+
+size_t AdditionalFieldList::size() const
+{
+    return m_additional_fields.size();
+}
+
+bool AdditionalFieldList::empty() const
+{
+    return m_additional_fields.empty();
+}
+
+bool AdditionalFieldList::WellFormed(const PollType poll_type) const
```

Comment:
Lacking builder-side checks here.

Reply:
Not sure I know what you mean by this. I will go look.

Reply:
Ok. I think this is fixed.